Sample size calculation

selenabr opened the pull request:
## Description
Implementation of sample_size_calculation to calculate the required sample size for a selected protein. For this purpose, the variance method was added and an output field was implemented to display the result. The variance and the sample size calculation-method are tested with test-data in test_power_analysis.py

<!-- Description of the change made in the pull-request. -->

## Changes
<!-- Where and what about the code was changed. The most important files and changes should be listed here. -->

sample_size_calculation + variance method:
- protzilla/data_analysis/power_analysis.py
- ui/runs/forms/data_analysis.py
- protzilla/methods/data_analysis.py

Display output field:
- ui/runs/templates/runs/details.html
- ui/runs/views.py
- protzilla/steps.py

Test:
- tests/protzilla/data_analysis/test_power_analysis.py


 
**Mergeability**
- [ ] main-branch has been merged into local branch to resolve conflicts
- [ ] The tests and linter have passed AFTER local merge
- [ ] The code has been formatted with `black`
 
**Code review**
- [ ] I have self-reviewed my code.
- [ ] At least one other developer reviewed and approved the changes

selenabr committed:
added sample size calculation in methods\data_analysis.py and forms\data_analysis.py

selenabr committed:
enabled possibility to choose one protein for calculation dependent on significance from t-test

selenabr committed:
fixed errors with missing inputs

selenabr committed:
added variance calculation and testing function and edited sample size calculation function

selenabr committed:
fixed some errors

selenabr committed:
output field for result

selenabr committed:
Merge branch 'dev' into bachelor-thesis-selena

# Conflicts:
#	protzilla/methods/data_analysis.py
#	ui/runs/form_mapping.py
#	ui/runs/forms/data_analysis.py

selenabr committed:
further implementation of output field for result

selenabr committed:
display display_output in output field

selenabr committed:
display_output field displayed in the same size and position as the other fields

selenabr committed:
test function for sample_size_calculation

selenabr committed:
Merge branch 'dev' into bachelor-thesis-selena

# Conflicts:
#	protzilla/methods/data_analysis.py
#	ui/runs/forms/data_analysis.py

selenabr committed:
edited description of function

hendraet commented on protzilla/data_analysis/power_analysis.py:
```diff
@@ -0,0 +1,91 @@
+    """
+
+    if intensity_name is None:
+        intensity_name = "Normalised iBAQ"
```
Assumes that data has to be normalized before feeding into the step. Otherwise the column doesn't exist. I would say that is an unnecessary limitation that is not transparent to the user

hendraet replied:
I feel that this could just the default argument if it is set anyways. Or is there a reason why the default has to be `None`?

hendraet commented on protzilla/data_analysis/power_analysis.py:
```diff
@@ -0,0 +1,91 @@
+def sample_size_calculation(
+    differentially_expressed_proteins_df: pd.DataFrame,
+    significant_proteins_df: pd.DataFrame,
+    significant_proteins_only: bool,
```
Unused

hendraet commented on protzilla/data_analysis/power_analysis.py:
```diff
@@ -0,0 +1,91 @@
+    group2: str,
+    selected_protein_group: str,
+    intensity_name: str = None
+) -> float:
```
Adapt annotation

hendraet commented on ui/runs/forms/data_analysis.py:
```diff
@@ -1090,4 +1089,93 @@ def fill_form(self, run: Run) -> None:
+    def fill_form(self, run: Run) -> None:
+        self.fields["t_test_results"].choices = get_t_test_results(run)
+
+class PowerAnalysisSampleSizeCalculationForm(MethodForm):
```
In general, you should run a code formatter. Often there too few blank lines (between methods or here above the class, you should leave two empty lines) and unnecessary whitespaces around equal signs (as below in the `...Field()`

hendraet reviewed (commented):
Looks good so far. I have written pretty much the same code for the calculation :)

As usual, there are still minor things that could be improved, but that's mainly code style.

selenabr committed:
check if implemented function of Paper (Cairns et al., 2009) and library-function of Sample Size Calculation have the same result

selenabr committed:
power calculation and test of library-function and implemented paper-function

hendraet commented on protzilla/data_analysis/power_analysis.py:
```diff
@@ -0,0 +1,288 @@
+        intensity_name=intensity_name,
+    )
+    sample_size = differentially_expressed_proteins_df.groupby('Group')['Sample'].count()
+    z_beta = fc_threshold * np.sqrt(sample_size/(2*variance_protein_group**2))-z_alpha
```
```suggestion
    z_beta = fc_threshold * np.sqrt(sample_size / (2 * variance_protein_group)) - z_alpha
```

I think the square is too much since we are already dealing with variances and not standard deviations. (Also some minor formatting issues)

selenabr committed:
added test for power_calculation method

selenabr committed:
fixed constructor error

selenabr committed:
sample size calculation for different group sizes (Cohen 1988) and moved validation methods to separate file

selenabr committed:
code formatting, resolved comments (output not a float, significant_proteins_only, intensity_name)

selenabr committed:
feature: user can choose whether metadata contains a column for individuals. If so, the mean values per individual are used to calculate the power and sample size.

selenabr committed:
adapted test for power_calculation and sample_size_calculation and checked values from paper of Cairns

selenabr committed:
added function that calculates sample size for all proteins and shows the distribution in a violin plot

selenabr committed:
formatting

selenabr committed:
commented the dataframe-output-stuff out, otherwise violin plot couldn't be displayed anymore (WIP...)

selenabr committed:
changed color of violinplot and added axis-description

hendraet commented on protzilla/data_analysis/power_analysis.py:
```diff
@@ -0,0 +1,341 @@
+            **violin_plot_args
+        )
+    )
+    fig.update_layout(
```
If you don't add traces to a figure dynamically (e.g. in a for loop), you can also pass the trace directly to `go.Figure()`

hendraet commented on protzilla/data_analysis/power_analysis.py:
```diff
@@ -0,0 +1,341 @@
+        spanmode='hard',
+        span=[0, required_sample_size_for_all_proteins],
+        fillcolor='rgba(0,0,0,0)'
+    )
```
would pass these arguments directly to `go.Violin()`. Since you are not reusing them somewhere else, it just makes the code harder to read because these args are in a different place

hendraet commented on protzilla/data_analysis/power_analysis.py:
```diff
@@ -0,0 +1,341 @@
+
+    fig.add_trace(
+        go.Violin(
+            x=["Protein Groups"] * len(required_sample_sizes),
```
Would omit the `x` parameter and just use `name="Protein Groups"`. It's easier to read

selenabr committed:
changed color of violinplot and removed axis-description

selenabr committed:
resolved comments

selenabr committed:
Added function to get dataframes with sample size column as output

selenabr committed:
Added power_calculation_for_all_proteins to calculate minimum power for all proteins

selenabr committed:
Fixed hover display of violin plots

selenabr committed:
fixed typo and removed unnecessary comment

selenabr committed:
calculations for thesis (should be removed before merging into dev)

selenabr committed:
calculations for thesis (should be removed before merging into dev)

selenabr committed:
put calculation for thesis into comment and changed description of methods "...for All Proteins"

selenabr committed:
Add files via upload

meta file that includes an additional column that identifies the individual sample IDs.

selenabr committed:
Merge branch 'dev' into bachelor-thesis-selena

selenabr committed:
merge bachelor-thesis-selena into dev

selenabr committed:
fixed error in power_analysis.py (constants.color) and commented file test_power_analysis.py

krohnjonas commented on ui/runs/views.py:
```diff
@@ -159,6 +166,9 @@ def detail(request: HttpRequest, run_name: str):
             description=description,
             method_form=method_form,
             is_form_dynamic=method_form.is_dynamic,
+            plot_form=plot_form,
```
We merged the plot form with the calculate form so that every step owns only one form containing all input fields. So this line shouldn't be necessary anymore.

krohnjonas commented on user_data/workflows/overhaul.yaml:Zone.Identifier:
For what is this file? It doesn't look like a normal workflow and I can't figure out what's the purpose of this file.

krohnjonas replied:
Unfortunately `:` aren't allowed in paths on windows machines so that i can't checkout to your branch because of this file. I hope so much that you don't need this file :D

selenabr replied:
Oh, I also don't know what this file is for. I've asked the others from the old project, but nobody seems to know. Also, the git history is empty, so should I just delete it?

krohnjonas commented on tests/protzilla/data_analysis/test_power_analysis.py:
Why are these test commented out?

selenabr replied:
Actually, the first tests up to line 202 shouldn't be commented out. They tested the new methods on the old branch, and they worked. I think I commented them out because the methods didn't work on the dev branch due to the new changes...

Jonas0000 committed:
changed steps to new format

krohnjonas commented on user_data/workflows/standard.yaml:
Should the new steps really be part of the standard workflow?

selenabr replied:
I'm not sure, maybe we should talk to Chris about this. But I think it's totally fine if the new steps are just available in PROTzilla :)

Jonas0000 reviewed (commented):
Ich habe deinen PR einmal gereviewed, damit der möglichst bald gemerged werden kann. Sehr cool, dass du so viele neue Funktionen eingebaut hast!
Ein paar kleinere Fragen habe ich dir an den Code geschrieben.
Außerdem werde ich gleich mal Änderungen commiten, mit denen deinen Steps an den neuen Syntax angepasst werden. Inhaltlich habe ich mir deine neuen Steps nicht angeschaut.

Ich habe gesehen, dass du sehr viel Code geformatted hast - vermutlich automatisch durch einen formatter? Grundsätzlich finde ich das sehr ut und es macht den Code deutlich lesbarer.
Wäre es aber für dich einfach möglich die Formatänderungen rückgängig zu machen?
Ich glaube das würde es uns deutlich einfacher machen, den Code ins neue Protzilla zu mergen. Wenn nicht, bekommen wir das bestimmt auch so hin. Ich frage hierzu auch nochmal im BP nach, wie dort die Meinung ist.

selenabr reviewed (commented):
Vielen Dank fürs Reviewen!  :) 
Bezüglich des formatting: Wir haben alle den Black formatter benutzt. Ich weiß nicht, ob es möglich ist, alle Formatierungsänderungen rückgängig zu machen. Vielleicht könnt ihr einfach über den branch das Formatting rüberlaufen lassen, was ihr selbst benutzt, falls es nicht black ist?
